src/optimizer/interpreter.ts

```diff
@@ -870,18 +882,42 @@ export class Interpreter {
     }
 
     public interpretName(ast: A.AstId): A.AstLiteral {
-        if (hasStaticConstant(this.context, idText(ast))) {
-            const constant = getStaticConstant(this.context, idText(ast));
+        const name = idText(ast);
+
+        if (hasStaticConstant(this.context, name)) {
+            const constant = getStaticConstant(this.context, name);
             if (constant.value !== undefined) {
                 return constant.value;
-            } else {
+            }
+
+            // Since we call `interpretExpression` on a constant value below, we don't want
+            // infinite recursion due to circular dependencies. To prevent this, let's collect
+            // all the constants we process in this iteration. That way, any circular dependencies
+            // will result in a second occurrence here and thus an early (before stack overflow)
+            // exception being thrown here.
+            if (this.visitedConstants.has(name)) {
                 throwErrorConstEval(
-                    `cannot evaluate declared constant ${idTextErr(ast)} as it does not have a body`,
+                    `cannot evaluate ${name} as it has circular dependencies: [${this.constantComputationPath.join(" -> ")} -> ${name}]`,
                     ast.loc,
                 );
             }
+            this.visitedConstants.add(name);
+
+            if (constant.ast.kind === "constant_def") {
+                this.constantComputationPath.push(name);
+                constant.value = this.interpretExpression(
+                    constant.ast.initializer,
+                );
+                this.constantComputationPath.pop();
```

Comment:
Since these `push` / `pop` operations need to be always balanced, let's define a helper that ensures that

Reply:
Done